File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/OrdersRevenue.scala
package io.devorbit.sparklabs.batch.sales

import io.devorbit.sparklabs.batch.sales.Utils
import org.apache.spark.sql.functions._

object OrdersRevenue extends App {

  Utils.orderItems.join(Utils.products, Seq("product_id"), "inner")
    .withColumn("revenue", col("price") * col("quantity"))
    .groupBy(col("order_id"))
    .agg(sum("revenue").as("total_revenue"), count("*").as("total_orders"))
    .orderBy(desc("total_revenue"))
    .show(false)

//  Thread.sleep(1000000)
}

File: r/sparklabs/batch/sales/sql/orders_revenue.R
source("r/sparklabs/batch/sales/sql/utils.R")

showDF(run_sql("
  SELECT oi.order_id,
         SUM(p.price * oi.quantity) AS total_revenue,
         COUNT(*) AS total_orders
  FROM order_items oi
  JOIN products p ON oi.product_id = p.product_id
  GROUP BY oi.order_id
  ORDER BY total_revenue DESC
"), truncate = FALSE)
